examples/webgl_depth_peeling.html

```diff
@@ -0,0 +1,272 @@
+					renderer.setRenderTarget( layer );
+					renderer.clear();
+					renderer.render( depthPeelingScene, camera );
+					return layer.depthTexture;
+
+				}, one );
+
+				renderer.setRenderTarget( originalRenderTarget );
+				renderer.clear();
+
+				for ( const layer of layers ) {
+
+					quad.material.uniforms.tDiffuse.value =
+						layer.texture;
+					quad.material.needsUpdate = true;
+					quad.render( renderer );
+
+				}
+
+				renderer.autoClear = originalAutoClear;
+
+			}
+
+			init().then( animate );
+
+			async function init() {
+
+				renderer.setPixelRatio( window.devicePixelRatio );
+				renderer.setSize( window.innerWidth, window.innerHeight );
+				document.body.appendChild( renderer.domElement );
+
+				camera.position.z = 5;
+
+				renderer.shadowMap.enabled = true;
+
+				const dirLight = new THREE.DirectionalLight();
+				dirLight.position.set( 0, 3, 0 );
+				dirLight.castShadow = true;
+				scene.add( dirLight );
+				scene.add( new THREE.AmbientLight( undefined, 0.5 ) );
+
+				const sphere = new THREE.Mesh(
+					new THREE.SphereBufferGeometry(),
+					new THREE.MeshStandardMaterial()
+				);
+				sphere.translateY( 3 ).translateX( 1.5 );
+				sphere.castShadow = true;
+				scene.add( sphere );
+
+				const knot = new THREE.Mesh(
+					new THREE.TorusKnotBufferGeometry( undefined, undefined, 128, 32 ),
+					new THREE.MeshStandardMaterial( {
+						transparent: true,
+						opacity: 0.7,
+						side: THREE.DoubleSide,
+					} )
+				);
+				knot.receiveShadow = true;
+				scene.add( knot );
+
+				scene.add(
+					new THREE.Mesh(
+						new THREE.BoxBufferGeometry(),
+						new THREE.MeshBasicMaterial( { color: 0xf0a000 } )
+					)
+				);
+
+				const texture = await new THREE.TextureLoader().loadAsync(
+					'textures/sprite0.png'
+				);
+				const plane = new THREE.Mesh(
+					new THREE.PlaneBufferGeometry( 3, 3 ),
+					new THREE.MeshStandardMaterial( { map: texture, side: THREE.DoubleSide } )
+				);
+				plane.translateX( - 1.6 );
+				plane.translateY( 1.5 );
+				scene.add( plane );
+
+				const plane2 = new THREE.Mesh(
+					new THREE.PlaneBufferGeometry( 3, 3 ),
+					new THREE.MeshStandardMaterial( {
+						map: texture,
+						transparent: true,
+						side: THREE.DoubleSide,
+					} )
+				);
+				plane2
+					.translateX( - 1.2 )
+					.translateY( - 1.5 )
+					.translateZ( 0 )
+					.rotateY( - 2 * Math.PI * ( 1 / 10 ) );
+				scene.add( plane2 );
+
+				depthPeelingScene = scene.clone( true );
+
+				resizeLayers( window.innerWidth, window.innerHeight, renderer.getPixelRatio() );
+				resizeDepth( window.innerWidth, window.innerHeight, renderer.getPixelRatio(), AppState.Depth );
+
+				depthPeelingScene.traverse( ( obj ) => {
+
+					if ( obj instanceof THREE.Mesh && obj.material instanceof THREE.Material ) {
+
+						const clonedMaterial = obj.material.clone();
+						clonedMaterial.blending = THREE.NoBlending;
+						clonedMaterial.onBeforeCompile = ( shader ) => {
```

Comment:
In the past, we've come to the conclusion to not add this kind of shader patching anymore since it easily breaks when built-in materials are changed. 

#15312 also did that and it always felt like a hack. But I assume there is no other way to support depth peeling right? Meaning the technique requires a change in the shader code, correct?

Reply:
Or I can make snapshot of current(r141) `MeshStandardMaterial` fragment shader and embed it using `ShaderMaterial`? This way it won't break by any shader update.

Reply:
Um, I'm not yet sure what's best in this context.

Reply:
[GLSL specification](https://www.khronos.org/registry/OpenGL/specs/gl/GLSLangSpec.4.60.pdf) (p.131) states

> The function main is used as the entry point to a shader executable. A shader need not contain a
> function named main, but one shader in a set of shaders linked together to form a single shader
> executable must, or a link-time error results. This function takes no arguments, returns no value,
> and must be declared as type void:

So I think matching `void main()` should be practically very safe.

But I understand your concerns so I'll leave it up to your decision.

Reply:
Hello @Mugen87. I can write a custom shader that supports basic lighting and shadowing independent from ThreeJS shader chunks (e.i no #include<...>) rather than using built-in materials.  Would you consider re-reviewing if I do that?

Reply:
Hmm probably not shadowing because it will lengthen the code (especially without shader chunking) and is not relevant to Depth Peeling itself 🤔

Reply:
> I can write a custom shader that supports basic lighting and shadowing independent of ThreeJS shader chunks (e.i no #include<...>) rather than using built-in materials.

I don't think that's a good idea since we want to keep custom shader code in the examples at a minimum. It's probably better to think about an integration in the renderer.

Reply:
@mrdoob What do you think about this? OIT has been discussed several times (#9977) and there are now two PRs trying to implement an example for this. However, both (this one and #15312) require shader modifications to make things work. This kind of built-in material modification turned out to be problematic in the past and we have eventually removed related examples (e.g. box-projected env maps, instanced rendering).

Nevertheless, do you think a depth peeling example would be still beneficial? Or should we think about an integration in the renderer?

Reply:
I understood. Thank you for the review though. 👍

Reply:
Depth Peeling is analogous to shadowing. It requires two depth-comparison (one regular depth-test, one "peeling test"). So not only #15312 and mine but every depth peeling examples will require shader modification.

Reply:
I still think `void main() {` match should be practically safe since those four tokens will always exist consecutively according to GLSL specification. 
But I understand it's a "hack" nevertheless and your reluctance to this approach.

Reply:
Depth Peeling is simple technique yet there are not many examples out there especially with THREE.JS. So I think it will have benefits at least to some extent for users.

Reply:
I think what @Mugen87 is suggesting is to add the depth peeling to the renderer itself rather than to create an example or class.

Reply:
@LeviPesin Yes, I understood that part. But he asked @mrdoob second opinion and I pryingly added mine 😅

Reply:
All in all, please close the PR if you guys think it's against the THREE.JS's policy. I'm happy either way 👍